File: src/doc_update_tool/file_types/excel_adapter.py
"""openpyxl-backed adapter for zh/en paired columns in an Excel sheet.

Column layout (which letters hold zh vs en, which columns form a row key) is entirely
driven by ExcelMapping — nothing here assumes a specific customer layout. Use
`docmgr inspect-excel` to discover the right column letters for a new document.
"""

from __future__ import annotations

from pathlib import Path

import openpyxl
from openpyxl.workbook import Workbook

from doc_update_tool.config import ExcelMapping
from doc_update_tool.models import Segment, SegmentUpdate


class ExcelAdapter:
    """Loads/saves workbooks in place so styles, merged cells, formulas and unrelated
    sheets are never touched — only the configured en_columns cells are written to."""

    def __init__(self, mapping: ExcelMapping):
        self.mapping = mapping

    def load(self, path: Path) -> Workbook:
        return openpyxl.load_workbook(path)

    def _sheet(self, wb: Workbook):
        if self.mapping.sheet_name == "auto":
            return wb.active
        return wb[self.mapping.sheet_name]

    def extract_segments(self, wb: Workbook) -> list[Segment]:
        ws = self._sheet(wb)
        header_row = self.mapping.header_row
        headers = {col: ws[f"{col}{header_row}"].value for col in self.mapping.zh_columns}

        segments: list[Segment] = []
        for row in range(header_row + 1, ws.max_row + 1):
            zh_values = {col: ws[f"{col}{row}"].value for col in self.mapping.zh_columns}

            row_key = None
            if self.mapping.key_columns:
                key_parts = [str(ws[f"{col}{row}"].value or "").strip() for col in self.mapping.key_columns]
                row_key = "|".join(key_parts) if any(key_parts) else None

            row_has_content = any(
                v is not None and str(v).strip() != "" for v in zh_values.values()
            )
            if not row_has_content and row_key is None:
                continue  # fully blank row, likely trailing whitespace in the sheet

            for zh_col, en_col in zip(self.mapping.zh_columns, self.mapping.en_columns):
                zh_text = zh_values[zh_col]
                en_text = ws[f"{en_col}{row}"].value
                segments.append(
                    Segment(
                        location_id=f"{ws.title}!{zh_col}{row}",
                        en_location_id=f"{ws.title}!{en_col}{row}",
                        row_key=row_key,
                        field_name=headers.get(zh_col),
                        zh_text=str(zh_text) if zh_text is not None else "",
                        en_text=str(en_text) if en_text is not None else "",
                    )
                )
        return segments

    def apply_translations(self, wb: Workbook, updates: list[SegmentUpdate]) -> None:
        for update in updates:
            sheet_name, cell_ref = update.en_location_id.split("!", 1)
            wb[sheet_name][cell_ref] = update.en_text

    def save(self, wb: Workbook, path: Path) -> None:
        wb.save(path)


def inspect(path: Path, sheet_name: str = "auto", header_row: int = 1, sample_rows: int = 3) -> list[dict]:
    """Summarize each column's header + sample values, to help pick zh_columns/en_columns
    for a customer config without knowing the layout up front. Flags formula cells since
    the adapter assumes zh/en columns hold plain text."""
    wb = openpyxl.load_workbook(path, data_only=False)
    ws = wb.active if sheet_name == "auto" else wb[sheet_name]

    result = []
    for col_idx in range(1, ws.max_column + 1):
        col_letter = ws.cell(row=1, column=col_idx).column_letter
        header = ws.cell(row=header_row, column=col_idx).value
        samples = []
        has_formula = False
        for row in range(header_row + 1, min(header_row + 1 + sample_rows, ws.max_row + 1)):
            cell = ws.cell(row=row, column=col_idx)
            if isinstance(cell.value, str) and cell.value.startswith("="):
                has_formula = True
            samples.append(cell.value)
        result.append(
            {
                "column": col_letter,
                "header": header,
                "samples": samples,
                "has_formula": has_formula,
            }
        )
    return result

File: web/src/excel-adapter.ts
/** Mirrors file_types/excel_adapter.py, backed by ExcelJS instead of openpyxl.
 *
 * Real customer workbooks have multiple tabs with genuinely different layouts, so a
 * mapping is applied per worksheet name (Record<sheetName, SheetMapping>) rather than
 * once for "the" sheet — every configured sheet that exists in the workbook gets its
 * own segments extracted, and they're merged into one flat list (each Segment's
 * location already carries its sheet name, so diff reports/translation tables can
 * still tell which tab a change came from).
 *
 * The ExcelJS namespace is passed in rather than imported at the module level: in the
 * browser it's loaded as a global via <script src="libs/exceljs.min.js"> (this tool
 * ships zero bundled copies of exceljs so the static files stay small and cacheable),
 * while unit tests running under Node inject the real npm package. Only `import type`
 * is used below, which esbuild/tsc erase completely, so no bundler ever needs to
 * actually resolve the "exceljs" module at build time.
 */

import type ExcelJSNS from "exceljs";

import { cellText } from "./cell-text.ts";
import type { SheetMapping } from "./config.ts";
import type { Segment, SegmentUpdate } from "./models.ts";
import { toArrayBuffer } from "./xlsx-buffer.ts";

export class ExcelAdapter {
  private readonly ExcelJS: typeof ExcelJSNS;
  private readonly sheetMappings: Record<string, SheetMapping>;

  constructor(ExcelJS: typeof ExcelJSNS, sheetMappings: Record<string, SheetMapping>) {
    this.ExcelJS = ExcelJS;
    this.sheetMappings = sheetMappings;
  }

  async load(bytes: ArrayBuffer): Promise<ExcelJSNS.Workbook> {
    const wb = new this.ExcelJS.Workbook();
    await wb.xlsx.load(bytes);
    return wb;
  }

  /** `referenceWb` is only used as a fallback when `wb` doesn't have a sheet with the
   * mapping's exact name — real customer workbooks sometimes rename tabs between
   * versions (e.g. stripping an English suffix) while keeping the same tab order, so
   * this falls back to matching by position within `referenceWb` (typically the newly
   * uploaded document, since mappings are configured against its sheet names) rather
   * than silently dropping the whole sheet from the diff. */
  extractSegments(wb: ExcelJSNS.Workbook, referenceWb?: ExcelJSNS.Workbook): Segment[] {
    const segments: Segment[] = [];
    for (const mapping of Object.values(this.sheetMappings)) {
      if (mapping.mode !== "table") continue; // freeform sheets go through diffFreeformSheet instead
      const ws = resolveWorksheet(wb, mapping.sheetName, referenceWb);
      if (!ws) continue; // this workbook doesn't have that tab (even by position) — skip rather than fail
      segments.push(...extractFromSheet(ws, mapping));
    }
    return segments;
  }

  applyTranslations(wb: ExcelJSNS.Workbook, updates: SegmentUpdate[]): void {
    for (const update of updates) {
      const sepIdx = update.enLocationId.indexOf("!");
      const sheetName = update.enLocationId.slice(0, sepIdx);
      const cellRef = update.enLocationId.slice(sepIdx + 1);
      const ws = wb.getWorksheet(sheetName);
      if (!ws) throw new Error(`找不到工作表: ${sheetName}`);
      ws.getCell(cellRef).value = update.enText;
    }
  }

  async save(wb: ExcelJSNS.Workbook): Promise<ArrayBuffer> {
    const buffer = await wb.xlsx.writeBuffer();
    return toArrayBuffer(buffer as ArrayBuffer | ArrayBufferView);
  }
}

/** `referenceWb` is only used as a fallback when `wb` doesn't have a sheet with this
 * exact name — real customer workbooks sometimes rename tabs between versions (e.g.
 * stripping an English suffix) while keeping the same tab order, so this falls back to
 * matching by position within `referenceWb` (typically the newly uploaded document,
 * since mappings/freeform sheet choices are made against its sheet names) rather than
 * silently dropping the whole sheet from the diff. Exported so the freeform diff path
 * in app.ts (which isn't driven by ExcelAdapter) can resolve sheets the same way. */
export function resolveWorksheet(
  wb: ExcelJSNS.Workbook,
  sheetName: string,
  referenceWb?: ExcelJSNS.Workbook
): ExcelJSNS.Worksheet | undefined {
  const exact = wb.getWorksheet(sheetName);
  if (exact) return exact;
  if (!referenceWb) return undefined;
  const refIndex = referenceWb.worksheets.findIndex((w) => w.name === sheetName);
  if (refIndex === -1) return undefined;
  return wb.worksheets[refIndex];
}

function extractFromSheet(ws: ExcelJSNS.Worksheet, mapping: SheetMapping): Segment[] {
  const headerRow = mapping.headerRow;
  const headers = new Map<string, unknown>();
  for (const col of mapping.zhColumns) {
    headers.set(col, ws.getCell(`${col}${headerRow}`).value);
  }

  const segments: Segment[] = [];
  const lastRow = lastNonBlankRow(ws, headerRow, mapping);

  for (let row = headerRow + 1; row <= lastRow; row++) {
    const zhValues = new Map<string, unknown>();
    for (const col of mapping.zhColumns) {
      zhValues.set(col, ws.getCell(`${col}${row}`).value);
    }

    let rowKey: string | null = null;
    if (mapping.keyColumns.length > 0) {
      const keyParts = mapping.keyColumns.map((col) => cellText(ws.getCell(`${col}${row}`).value).trim());
      rowKey = keyParts.some((p) => p !== "") ? keyParts.join("|") : null;
    }

    const rowHasContent = [...zhValues.values()].some((v) => cellText(v).trim() !== "");
    if (!rowHasContent && rowKey === null) continue; // fully blank row

    for (let i = 0; i < mapping.zhColumns.length; i++) {
      const zhCol = mapping.zhColumns[i];
      const enCol = mapping.enColumns[i];
      segments.push({
        locationId: `${ws.name}!${zhCol}${row}`,
        enLocationId: `${ws.name}!${enCol}${row}`,
        rowKey,
        fieldName: headers.get(zhCol) != null ? cellText(headers.get(zhCol)) : null,
        zhText: cellText(zhValues.get(zhCol)),
        enText: cellText(ws.getCell(`${enCol}${row}`).value),
      });
    }
  }
  return segments;
}

/** ExcelJS's worksheet.rowCount can include trailing rows with only style/no value, so
 * we scan for the last row that actually has content in a mapped column instead of
 * trusting it blindly — mirrors the Python adapter iterating to ws.max_row and letting
 * the blank-row check drop trailing whitespace rows. */
function lastNonBlankRow(ws: ExcelJSNS.Worksheet, headerRow: number, mapping: SheetMapping): number {
  const candidateCols = [...mapping.zhColumns, ...mapping.keyColumns];
  let last = headerRow;
  const scanLimit = Math.max(ws.rowCount, ws.actualRowCount, headerRow);
  for (let row = headerRow + 1; row <= scanLimit; row++) {
    const hasContent = candidateCols.some((col) => cellText(ws.getCell(`${col}${row}`).value).trim() !== "");
    if (hasContent) last = row;
  }
  return last;
}

export interface InspectColumn {
  column: string;
  header: unknown;
  samples: unknown[];
  hasFormula: boolean;
}

export interface InspectSheet {
  sheetName: string;
  columns: InspectColumn[];
}

/** Lists every worksheet's columns (header + sample values), used to power the
 * point-and-click mapping UI instead of asking users to type raw column letters. */
export async function inspectWorkbook(
  ExcelJS: typeof ExcelJSNS,
  bytes: ArrayBuffer,
  opts: { headerRow?: number; sampleRows?: number } = {}
): Promise<InspectSheet[]> {
  const headerRow = opts.headerRow ?? 1;
  const sampleRows = opts.sampleRows ?? 3;

  const wb = new ExcelJS.Workbook();
  await wb.xlsx.load(bytes);

  return wb.worksheets.map((ws) => {
    const columns: InspectColumn[] = [];
    for (let colIdx = 1; colIdx <= ws.columnCount; colIdx++) {
      const columnLetter = ws.getColumn(colIdx).letter;
      const header = ws.getRow(headerRow).getCell(colIdx).value;
      const samples: unknown[] = [];
      let hasFormula = false;
      for (let row = headerRow + 1; row < headerRow + 1 + sampleRows; row++) {
        const cell = ws.getRow(row).getCell(colIdx);
        if (cell.type === ExcelJS.ValueType.Formula) hasFormula = true;
        samples.push(cell.value);
      }
      columns.push({ column: columnLetter, header, samples, hasFormula });
    }
    return { sheetName: ws.name, columns };
  });
}
